ceno_zkvm/src/witness.rs

```diff
@@ -130,10 +131,59 @@ impl<F: Sync + Send + Copy> Index<usize> for RowMajorMatrix<F> {
 /// struct is cloneable, for internallly it use Arc so the clone will be low cost
 #[derive(Clone, Default, Debug)]
 #[allow(clippy::type_complexity)]
-pub struct LkMultiplicity {
-    multiplicity: Arc<ThreadLocal<RefCell<[HashMap<u64, usize>; mem::variant_count::<ROMType>()]>>>,
+pub struct LkMultiplicityRaw<K: Copy + Clone + Debug + Eq + Hash + Send> {
+    multiplicity: Arc<ThreadLocal<RefCell<[HashMap<K, usize>; mem::variant_count::<ROMType>()]>>>,
+}
+
+impl<K: Copy + Clone + Debug + Eq + Hash + Send> LkMultiplicityRaw<K> {
+    /// Merge result from multiple thread local to single result.
+    pub fn into_finalize_result(self) -> [HashMap<K, usize>; mem::variant_count::<ROMType>()] {
```

Comment:
These are just for-loops at their core.  No need to pretend we are more functional than we are:

```rust
    /// Merge result from multiple thread local to single result.
    pub fn into_finalize_result(self) -> [HashMap<K, usize>; mem::variant_count::<ROMType>()] {
        let mut results = array::from_fn(|_| HashMap::new());
        for y in Arc::try_unwrap(self.multiplicity).unwrap() {
            for (result, addition) in izip!(&mut results, y.into_inner()) {
                for (key, value) in addition {
                    *result.entry(key).or_insert(0) += value;
                }
            }
        }
        results
    }
```